src/main/java/hudson/remoting/PingFailureAnalyzer.java

```diff
@@ -0,0 +1,225 @@
+package hudson.remoting;
+
+import org.jvnet.animal_sniffer.IgnoreJRERequirement;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.MonitorInfo;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+/**
+ * Generates diagnosis information (threaddump + stacktrace) when the pingThread fails
+ *
+ * <p>
+ * Useful to troubleshoot the slave disconnection failures on the agent side.
+ *
+ * @since 2.55
+ */
+public class PingFailureAnalyzer {
+
+    /**
+     * Channel we are going to study
+     */
+    private Channel c;
+
+    /**
+     * The cause of the ping failure
+     */
+    private Throwable cause;
+
+
+    /**
+     * Maximum of files to store inside pingFailureAnalyzer/thread-dumps and pingFailureAnalyzer/stacktraces.
+     * When the limit is reached out it starts overriding.
+     */
+    private final static int MAX_DIAGNOSIS_FILES = 5;
+
+    /**
+     * Folder to store the thread dumps. Maintains most recent N files in a directory in cooperation with the writer.
+     */
+    private final PingThreadFileListCap threadDumps = new PingThreadFileListCap(new File("pingFailureAnalyzer/thread-dumps"), MAX_DIAGNOSIS_FILES);
+
+    /**
+     * Folder to store the stacktraces. Maintains most recent N files in a directory in cooperation with the writer.
+     */
+    private final PingThreadFileListCap stacktraces = new PingThreadFileListCap(new File("pingFailureAnalyzer/stacktraces"), MAX_DIAGNOSIS_FILES);
+
+    public PingFailureAnalyzer(Channel c, Throwable cause) {
+        this.c = c;
+        this.cause = cause;
+    }
+
+
+    /**
+     * Save the stacktrace which produces the ping failure into pingFailureAnalyzer/stacktraces
+     *
+     * @param cause
+     *      The Throwable
+     */
+    public void saveStackTrace(Throwable cause) {
+        long iota = System.currentTimeMillis();
+        final SimpleDateFormat format = new SimpleDateFormat("yyyyMMdd-HHmmss.SSS");
+        File file = stacktraces.file(format.format(new Date(iota++)) + ".txt");
+        stacktraces.add(file);
+        PrintWriter writer = null;
+        try {
+            writer = new PrintWriter(file, "UTF-8");
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
+        cause.printStackTrace(writer);
+        writer.close();
+    }
+
+    /**
+     * Dumps all of the threads' current information to an output stream.
+     *
+     */
+    @IgnoreJRERequirement
```

Comment:
this seems to be wrong - why are you ignoring animal sniffer results - and what was the cause of the check failing?